feat(lite): introduce PWA manifest

MrGrymReaper opened the pull request:
### Description

This implements the initial PWA manifest for XO Lite. I requested this several months to a year ago, so decided to do it myself in the end.

### Checklist

- Commit
  - Title follows [commit conventions](https://bit.ly/commit-conventions)
  - Reference the relevant issue (`Fixes #007`, `See xoa-support#42`, `See https://...`)
  - If bug fix, add `Introduced by`
- Changelog
  - If visible by XOA users, add changelog entry
  - Update "Packages to release" in `CHANGELOG.unreleased.md`
- PR
  - If UI changes, add screenshots
  - If not finished or not tested, open as _Draft_

MrGrymReaper committed:
Create manifest.json

The initial PWA manifest file.

MrGrymReaper committed:
Update index.html

Adds the reference to the initial PWA manifest.

MrGrymReaper committed:
Update manifest.json

Corrected a typo.

MrGrymReaper committed:
Update CHANGELOG.md

Add the changelog entry.

ByScripts commented on @xen-orchestra/lite/manifest.json:
This file must be moved to the `public` directory.

MrGrymReaper replied:
Have made that change.

ByScripts commented on @xen-orchestra/lite/manifest.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "XO Lite",
+  "short_name": "XOLite",
+  "start_url": "/index.html",
```
```suggestion
  "start_url": "/",
```

MrGrymReaper replied:
Have made that change.

ByScripts commented on @xen-orchestra/lite/manifest.json:
```diff
@@ -0,0 +1,14 @@
+  "description": "Integrated local and lightweight solution to manage a local XCP-ng pool.",
+  "icons": [
+    {
+      "src": "/src/assets/favicon.svg",
```
Public files should be in the `public` directory.

`/src/assets/favicon.svg` should be copied to `/public` directory to be used.

```suggestion
      "src": "favicon.svg",
```

MrGrymReaper replied:
Have made that change.

ByScripts commented on @xen-orchestra/lite/manifest.json:
```diff
@@ -0,0 +1,14 @@
+    {
+      "src": "/src/assets/favicon.svg",
+      "sizes": "any",
+      "type": "image/svg"
```
```suggestion
      "type": "image/svg+xml"
```

MrGrymReaper replied:
Have made that change.

ByScripts commented on @xen-orchestra/lite/index.html:
```diff
@@ -3,6 +3,7 @@
   <head>
     <meta charset="UTF-8" />
     <link rel="icon" href="/favicon.ico" />
+    <link rel="manifest" href="/manifest.json">
```
Although `.json` extension is supported, it would be preferable to use the `.webmanifest` [recommended file extension](https://www.w3.org/TR/appmanifest/#media-type-registration)

```suggestion
    <link rel="manifest" href="/manifest.webmanifest">
```

MrGrymReaper replied:
Have already made that change.

ByScripts commented on @xen-orchestra/lite/CHANGELOG.md:
```diff
@@ -3,6 +3,7 @@
 ## **next**
 
 - Fix Typescript typings errors when running `yarn type-check` command (PR [#7278](https://github.com/vatesfr/xen-orchestra/pull/7278))
+- Introduce PWA Json Manifest.
```
```suggestion
- Introduce PWA Json Manifest (PR [#7291](https://github.com/vatesfr/xen-orchestra/pull/7291))
```

MrGrymReaper replied:
Applied the change.

ByScripts reviewed (changes requested):
Thanks a lot for your contribution.

However, some issues need to be addressed before merging.

ByScripts commented:
Additionally, for reference, if we want to go further on this subject in the future, maybe we should take a look at this [Vite PWA Plugin](https://vite-pwa-org.netlify.app/)

MrGrymReaper committed:
Prepare the icon file for use

In order for the resource to be used needs to copied from the source assets folder to public folder. This commit will perform this action.

MrGrymReaper committed:
Rename manifest.json to manifest.webmanifest

It's for the recommended extension even though the one from the research I did from Mozilla's Developer Resources had it so it was json extension.

MrGrymReaper committed:
Update index.html

Altered as requested and to maintain the reference manifest operating.

MrGrymReaper committed:
Create manifest.webmanifest

Relocated to the public directory and now named correctly.

MrGrymReaper committed:
Update manifest.webmanifest

Applied the change for the public path to svg icon.

MrGrymReaper committed:
Update @xen-orchestra/lite/CHANGELOG.md

To apply to the required standards.

Co-authored-by: Thierry Goettelmann <thierry@byscripts.info>

MrGrymReaper committed:
Update manifest.webmanifest

To apply the suggestion about the type for the svg icon image.

MrGrymReaper commented:
> Additionally, for reference, if we want to go further on this subject in the future, maybe we should take a look at this [Vite PWA Plugin](https://vite-pwa-org.netlify.app/)

That requires TypeScript as part of the plugin, I currently don't currently know this! So someone who is familiar with TypeScript would need to implement this.

MrGrymReaper committed:
Merge branch 'master' into xolitemanifest

ByScripts commented on @xen-orchestra/lite/index.html:
```diff
@@ -3,6 +3,7 @@
   <head>
     <meta charset="UTF-8" />
     <link rel="icon" href="/favicon.ico" />
+    <link rel="manifest" href="/public/manifest.webmanifest">
```
```suggestion
    <link rel="manifest" href="/manifest.webmanifest">
```

ByScripts commented on @xen-orchestra/lite/manifest.webmanifest:
Delete this file

ByScripts commented on @xen-orchestra/lite/public/manifest.webmanifest:
```diff
@@ -0,0 +1,14 @@
+  "description": "Integrated local and lightweight solution to manage a local XCP-ng pool.",
+  "icons": [
+    {
+      "src": "/publc/favicon.svg",
```
```suggestion
      "src": "favicon.svg",
```

Also, be careful to typo (`publc`).

ByScripts reviewed (changes requested):
Don't worry, I haven't asked you to use the Vite plugin, it's just a reference if we need to go further on this subject in the future.

Also, take care to read the code suggestions carefully, because your modifications don't match my suggestions.

Don't hesitate to use the "Commit suggestion" and "Add suggestion to batch" GitHub's feature.

Finally, although we're sincerely grateful for the contributions, it would be advisable to test the code before making a PR, because due to errors, this iteration still doesn't work.

Thank you.

MrGrymReaper committed:
Update @xen-orchestra/lite/public/manifest.webmanifest

Corrected a typo.

Co-authored-by: Thierry Goettelmann <thierry@byscripts.info>

MrGrymReaper committed:
Delete @xen-orchestra/lite/manifest.webmanifest

No longer needed.

MrGrymReaper committed:
Update @xen-orchestra/lite/index.html

Applied suggestion.

Co-authored-by: Thierry Goettelmann <thierry@byscripts.info>

MrGrymReaper commented:
> Don't worry, I haven't asked you to use the Vite plugin, it's just a reference if we need to go further on this subject in the future.
> 
> Also, take care to read the code suggestions carefully, because your modifications don't match my suggestions.
> 
> Don't hesitate to use the "Commit suggestion" and "Add suggestion to batch" GitHub's feature.
> 
> Finally, although we're sincerely grateful for the contributions, it would be advisable to test the code before making a PR, because due to errors, this iteration still doesn't work.
> 
> Thank you.

The Vite plugin if you enable it would allow for it to run in the background if the user grants permission. So as a result it would be able to send notifications even when the window is closed. Additionally PWA takes up far less memory than a full web browser, especially when a full web browser includes all of the addons (unless granted permission for an addon).

Background operator allows for the capacity to badge an app icon (overlay) which gets applied, so something similar to when you get new mail on a smart phone, tablet, modern MacOS, Windows and/or Linux.

What errors do you get as I would need to compile from source the code for this. Could you invoke the CI/CD and send me an artefact or test package please?

ByScripts commented:
> What errors do you get as I would need to compile from source the code for this. Could you invoke the CI/CD and send me an artefact or test package please?

There is no need to build to test.

Just run `yarn dev` and see if the app in installable.

For example, in Brave / Chrome, I get this icon in the search bar: ![PWA installable icon](https://github.com/vatesfr/xen-orchestra/assets/19408/dd571523-4133-4c57-baed-d413cde9abab)

Furthermore, you can also open Dev Tools, go to Lighthouse tab, click on "Analyze page load" then look at the PWA result to see errors and/or suggestions.

MrGrymReaper committed:
Merge branch 'master' into xolitemanifest

MrGrymReaper commented:
> Don't worry, I haven't asked you to use the Vite plugin, it's just a reference if we need to go further on this subject in the future.
> 
> Also, take care to read the code suggestions carefully, because your modifications don't match my suggestions.
> 
> Don't hesitate to use the "Commit suggestion" and "Add suggestion to batch" GitHub's feature.
> 
> Finally, although we're sincerely grateful for the contributions, it would be advisable to test the code before making a PR, because due to errors, this iteration still doesn't work.
> 
> Thank you.

Thanks for the reassurance.

ByScripts committed:
Linting manifest.webmanifest

pdonias commented:
@MrGrymReaper Thanks for the contribution! I'm ok with merging this, but is there any point in having a PWA manifest without registering a PWA in the code?

ByScripts commented:
@pdonias The manifest alone seems to allow to install XO Lite as an app.

pdonias commented:
Ok, I didn't know that was possible. Fine by me, then!
